Add more details about Kubelet -> Node plugin communication

bart0sh opened the pull request:
Updated KEP with more details about Kubelet

bart0sh committed:
dynamic resource allocation: add new parameter to NodePrepareResource API

bart0sh committed:
dynamic resource allocation: update 'Extend Device Plugins' section

Explained possible addition of NodePrepareResource and
NodeUnprepareResource calls to the Device Plugins framework.

bart0sh committed:
dynamic resource allocation: update SLO

Updated Service Level Objectives section:
 explained possible performance impact for Kubelet operations.

bart0sh committed:
dynamic resource allocation: add kubelet flow diagram

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1332,8 +1335,9 @@ with “volume” replaced by “resource” and volume specific parts removed.
 that this RPC will be executed on the node where the resource will be
-used.  The Plugin SHALL return device name and kind for allocated
-device[s].
+used.  Resource Claim UID and allocation result attributes should be
```
On chat, we discussed that claim name/namespace/UID should be passed to identify the claim. They are useful for debugging and enable the resource driver to retrieve the full object, should that ever be needed (normally it shouldn't).

As it stands now, the KEP talks about a "Resource Claim UID" without defining what that UID is.

bart0sh replied:
Thanks for pointing out to this. Fixed.

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1724,6 +1730,14 @@ Pod scheduling and startup are more important. However, expected performance
 will depend on how resources are used (for example, how often new Pods are
 created), therefore it is impossible to predict what reasonable SLOs might be.
 
+There shouldn't be a significant performance impact on the Kubelet side as
+Resource manager component will do its work similarly to the Volume manager:
```
Arguing that it will have little impact because it is similar to some other component seems a bit optimistic to me :sweat_smile: 

That other component might have significant impact, and then so will we. I would rephrase this:
```suggestion
Resource manager component will do its work similarly to the Volume manager, but the overhead and complexity should be lower:
```

bart0sh replied:
The similarity to Volume manager is not the only argument. It's just a start of the explanation. It simply refers to the fact that both upper level components are run as goroutines, similarly to Volume manager.

pohly replied:
But it is not a valid argument... similarity doesn't imply anything regarding performance impact.

bart0sh replied:
It does in my opinion. If it would do it sequentially(i.e. just calling some functions, not a goroutines) it can slow down the Kubelet.

pohly replied:
That it uses goroutines is a valid argument.

Let's just simplify the statements, then it will become clearer, okay? Start by stating that the resource manager will be similar to the volume manager. Describe how it will be simpler. Then conclude with an estimate of the the performance overhead.

pohly replied:
Also beware that "performance" could mean many things. If we consider the time required to start a pod, then the resource manager is on the critical path and running its code in goroutines doesn't change anything.

pohly replied:
Claiming "no significant performance impact" without being able to provide benchmark results that back up such a claim is risky. Reviewers will jump on this and ask for proof.

bart0sh replied:
>  Reviewers will jump on this and ask for proof.

Can you suggest how to put it better?

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1724,6 +1730,14 @@ Pod scheduling and startup are more important. However, expected performance
+desired state of world populator) will run as goroutines. Resource
+preparation should be fairly quick as in most cases it simply
+creates CDI file 1-3 Kb in size. Unpreparing resource usually means
+deleting CDI file, so it should be quick as well.
```
Add:

```suggestion
deleting CDI file, so it should be quick as well.
The complexity is lower than in the volume manager
because there is only one global operation needed (prepare vs. 
attach + publish for each pod) and reconstruction
after a kubelet restart is simpler (call NodePrepareResource
again vs. trying to determine whether volumes
are mounted).
```

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1918,6 +1932,13 @@ currently declared on the Pod and Device Plugin level.
+developers of the Device Plugins would have to implement mandatory
+API calls (ListAndWatch, Allocate), which could create confusion
+as those calls are meaningless for the Dynamic Resource Allocation
+purposes.
```
```suggestion
purposes.

Even worse, existing device plugins would have to implement the new
calls with stubs that return errors because the generated Go interface
will require them.
```

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1918,6 +1932,13 @@ currently declared on the Pod and Device Plugin level.
 Extending the Device Plugins API to use [Container Device Interface](https://github.com/container-orchestrated-devices/container-device-interface)
 would help address some of the requirements, but not all of them.
 
+NodePrepareResource and NodeUnprepareResource could be added as optional
+calls to the Device Plugins API. However, this would mean that
```
Let's be clearer about what "optional calls" means:

```suggestion
NodePrepareResource and NodeUnprepareResource could be added to the Device Plugins API and only get called for
resource claims.

However, this would mean that
```

pohly reviewed (changes requested):
Regarding the diagram: can you turn it into a PlantUML diagram? See the other files in the directory. I can help if something with the syntax isn't clear. Then it'll be easier to discuss the content of it.

One thing I noticed is that "Pod terminates" is an unclear state. What exactly is it?

I think it should say "Pod has DeletionTimestamp and all containers are stopped" instead.

The Pod object then still exists. kubelet needs to clean up (the next steps in the diagram). The second last state then is "Pod object ready for removal", followed by a "kubelet sets DeletionGracePeriodSeconds to zero".

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh committed:
dynamic resource allocation: add new API arguments

Updated list of parameters for the NodePrepareResource and
NodeUnprepareResource API calls:
- added namespace and claim_name parameters
- changed parameter name resource_id -> claim_uuid

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1724,6 +1752,19 @@ Pod scheduling and startup are more important. However, expected performance
 will depend on how resources are used (for example, how often new Pods are
 created), therefore it is impossible to predict what reasonable SLOs might be.
 
+Resource manager component will do its work similarly to the Volume manager, but the overhead and complexity should be lower:
+both upper level components of the Resource manager (reconciler and
+desired state of world populator) will run as goroutines. Resource
+preparation should be fairly quick as in most cases it simply
+creates CDI file 1-3 Kb in size. Unpreparing resource usually means
+deleting CDI file, so it should be quick as well.
+The complexity is lower than in the volume manager
+because there is only one global operation needed (prepare vs. 
+attach + publish for each pod) and reconstruction
+after a kubelet restart is simpler (call NodePrepareResource
+again vs. trying to determine whether volumes
+are mounted).
```
```suggestion
The resource manager component will do its work similarly to the 
existing volume manager, but the overhead and complexity should
be lower:

* Resource preparation should be fairly quick as in most cases it simply
  creates CDI file 1-3 Kb in size. Unpreparing resource usually means
  deleting CDI file, so it should be quick as well.

* The complexity is lower than in the volume manager
  because there is only one global operation needed (prepare vs. 
  attach + publish for each pod).
  
* Reconstruction after a kubelet restart is simpler (call 
  NodePrepareResource again vs. trying to determine whether 
  volumes are mounted).
```

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1332,8 +1335,16 @@ with “volume” replaced by “resource” and volume specific parts removed.
+and enable the resource driver to retrieve the full ResourceClaim object,
+should that ever be needed (normally it shouldn't).
+
+The Plugin SHALL return device name and kind for allocated device[s].
```
```suggestion
The Plugin SHALL return fully qualified device name[s].
```

bart0sh commented:
> One thing I noticed is that "Pod terminates" is an unclear state. What exactly is it?
> 
> I think it should say "Pod has DeletionTimestamp and all containers are stopped" instead.
> 
> The Pod object then still exists. kubelet needs to clean up (the next steps in the diagram). The second last state then is "Pod object ready for removal", followed by a "kubelet sets DeletionGracePeriodSeconds to zero".

I don't think it's a good idea to use so detailed and implementation-specific details as DeletionGracePeriodSeconds.
What I tried to do is to refer to the last of 3 phases of pod lifecycle mentioned in the pod_workers.go:
```
// PodWorkType classifies the three phases of pod lifecycle - setup (sync),
// teardown of containers (terminating), cleanup (terminated).
type PodWorkType int

const (
	// SyncPodWork is when the pod is expected to be started and running.
	SyncPodWork PodWorkType = iota
	// TerminatingPodWork is when the pod is no longer being set up, but some
	// containers may be running and are being torn down.
	TerminatingPodWork
	// TerminatedPodWork indicates the pod is stopped, can have no more running
	// containers, and any foreground cleanup can be executed.
	TerminatedPodWork
)
```
Would it be ok if I use "Pod is stopped and all its containers are stopped" or "Pod terminated" instead of "Pod terminates"?

pohly commented:
Using "Pod terminated" is fine with me.

pohly commented:
> implementation-specific details as DeletionGracePeriodSeconds

Side note: this is part of the Kubernetes API. That kubelet does this is not an implementation detail, other components (like the generic ephemeral inline volume controller) rely on it.

bart0sh committed:
dynamic resource allocation: use PlantUML diagram

Used PlantUML for kubelet diagram.
Generated diagram image kubelet.png from the kubelet.puml

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh committed:
Update keps/sig-node/3063-dynamic-resource-allocation/README.md

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

bart0sh commented:
@pohly Thank you for the suggestions! Accepted all of them and used PlantUML for the kubelet diagram.

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1724,6 +1752,22 @@ Pod scheduling and startup are more important. However, expected performance
+
+* Resource preparation should be fairly quick as in most cases it simply
+  creates CDI file 1-3 Kb in size. Unpreparing resource usually means
+  deleting CDI file, so it should be quick as well.
```
Beware that this ignores that "real" resource drivers may have to do more work to update resource tracking. We don't need to highlight this here, but it may come up in the KEP review.

bart0sh replied:
I understand that. However, we stated in the KEP that recommended way is to do all heavy work on allocation stage. We can point reviewers to that, I believe.

pohly commented on keps/sig-node/3063-dynamic-resource-allocation/kubelet.puml:
```diff
@@ -0,0 +1,12 @@
+-->[ **Kubelet** passes device Ids to the CRI runtime\l **CRI runtime** uses CDI JSON file(s) to update OCI container config\l **CRI runtime** creates pod containers ] "All Pod containers created"
+-->[ ... ] "Pod runs"
+-->[ ... ] "Pod terminated"
+-->[ **Kubelet** calls NodeUnprepareResource\l **Resource plugin** removes CDI JSON file(s)\l **Kubelet** removes pod from ResourceClaim.ReservedFor ] (*)
```
Error handling is not covered here (which is okay, it is a simplification) but also not explained in the main text. This might be worth adding.

bart0sh replied:
ok, will do

pohly reviewed (approved):
Let me merge it and ping reviewers. If you want to address my nits below, feel free to open another PR in parallel.

hj-johannes-lee commented on keps/sig-node/3063-dynamic-resource-allocation/README.md:
```diff
@@ -1332,8 +1335,16 @@ with “volume” replaced by “resource” and volume specific parts removed.
-used.  The Plugin SHALL return device name and kind for allocated
-device[s].
+used.  ResourceClaim.meta.Namespace, ResourceClaim.meta.UID,
+ResourceClaim.Name, ResourceClaim.meta.Namespace and
```
@bart0sh @pohly Sorry for seeing it after it's already merged, but there is "ResourceClaim.meta.Namespace" two times.
